File: common/src/main/java/StadiumBounce.java
import com.fasterxml.jackson.annotation.JsonProperty;

public class StadiumBounce {

    public static Rs noBounced() {
        return new Rs(null, null);
    }

    public static Rs collided(Point pos, Point vec) {
        return new Rs(null, new Collision(pos, vec));
    }

    public static Rs scored(String team) {
        return new Rs(team, null);
    }


    record Rq(@JsonProperty("seg") Segment seg) implements Jsonisable {}

    record Rs(
            @JsonProperty("scoredTeam") String scoredTeam,
            @JsonProperty("collision") Collision collision
    ) implements Jsonisable {
        boolean didBounce() {
            return scoredTeam != null || collision != null;
        }
    }

    record Collision(
            @JsonProperty("pos") Point pos,
            @JsonProperty("vec") Point vec
    ) implements Jsonisable {}
}

File: common/src/main/kotlin/demo/mesharena/common/StadiumBounce.kt
package demo.mesharena.common

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty


class StadiumBounce {
  data class Rq @JsonCreator constructor(
          @JsonProperty("seg") val seg: Segment
  ) : Jsonisable

  data class Rs @JsonCreator constructor(
          @JsonProperty("scoredTeam") val scoredTeam: String?,
          @JsonProperty("collision") val collision: Collision?
  ) : Jsonisable {
    fun didBounce(): Boolean {
      return scoredTeam != null || collision != null
    }
  }

  data class Collision @JsonCreator constructor(
          @JsonProperty("pos") val pos: Point,
          @JsonProperty("vec") val vec: Point
  ) : Jsonisable

  companion object {
    fun noBounced(): Rs {
      return Rs(null, null)
    }

    fun collided(pos: Point, vec: Point): Rs {
      return Rs(null, Collision(pos, vec))
    }

    fun scored(team: String): Rs {
      return Rs(team, null)
    }
  }
}
